### PR TITLE
Enable-Map-dialog

### PR DESCRIPTION
**Pull request by jubitjohn**
## Description

Option to enable Maps added to Preferences

### CODE REVIEW

**Review by ua741 — APPROVED** [review]

**Review by ua741 — APPROVED** [review]

File: apps/photos/public/locales/en/translation.json [hunk 1]
```diff
@@ -232,6 +232,15 @@
     "CAPTION_PLACEHOLDER": "Add a description",
     "LOCATION": "Location",
     "SHOW_ON_MAP": "View on OpenStreetMap",
+    "MAP":"Map",
+    "MAP_SETTINGS":"Map Settings",
+    "ENABLE_MAPS":"Enable Maps?",
+    "DISABLE_MAPS":"Disable Maps?",
+    "ENABLE_MAP_DESCRIPTION_1":"This will show your photos on a worldmap.",
```
Comment (abhinavkgrd) [inline apps/photos/public/locales/en/translation.json:239]: space between world and map ?

File: apps/photos/public/locales/en/translation.json [hunk 2]
```diff
@@ -232,6 +232,15 @@
...
+    "ENABLE_MAP_DESCRIPTION_2":"The map is hosted by OpenStreetMap, and the exact locations of your photos are never shared.",
```
Comment (abhinavkgrd) [inline apps/photos/public/locales/en/translation.json:240]: maybe a link to OpenStreetMap ?

File: apps/photos/src/components/Sidebar/Preferences/index.tsx [hunk 3]
```diff
@@ -7,15 +7,21 @@ import { useState } from 'react';
...
 
 export default function Preferences({ open, onClose, onRootClose }) {
     const [advancedSettingsView, setAdvancedSettingsView] = useState(false);
+    const [advancedMapSettingsView, setAdvancedMapSettingsView] =
+        useState(false);
 
     const openAdvancedSettings = () => setAdvancedSettingsView(true);
     const closeAdvancedSettings = () => setAdvancedSettingsView(false);
 
+    const openAdvancedMapSettings = () => setAdvancedMapSettingsView(true);
```
Comment (abhinavkgrd) [inline apps/photos/src/components/Sidebar/Preferences/index.tsx:22]: `openMapSettings` can be a better name

File: apps/photos/src/components/Sidebar/Preferences/index.tsx [hunk 4]
```diff
@@ -7,15 +7,21 @@ import { useState } from 'react';
 import { t } from 'i18next';
 
 import AdvancedSettings from '../AdvancedSettings';
+import AdvancedMapSettings from '../AdvancedMapSetting';
 import { LanguageSelector } from './LanguageSelector';
 import { EnteMenuItem } from 'components/Menu/EnteMenuItem';
 
 export default function Preferences({ open, onClose, onRootClose }) {
     const [advancedSettingsView, setAdvancedSettingsView] = useState(false);
+    const [advancedMapSettingsView, setAdvancedMapSettingsView] =
```
Comment (abhinavkgrd) [inline apps/photos/src/components/Sidebar/Preferences/index.tsx:16]: `mapSetting`

File: apps/photos/src/components/Sidebar/Preferences/index.tsx [hunk 5]
```diff
@@ -53,6 +59,11 @@ export default function Preferences({ open, onClose, onRootClose }) {
                                 label={t('ADVANCED')}
                             />
                         )}
+                        <EnteMenuItem
+                            onClick={openAdvancedMapSettings}
+                            endIcon={<ChevronRight />}
+                            label={t('Map')}
```
Comment (abhinavkgrd) [inline apps/photos/src/components/Sidebar/Preferences/index.tsx:65]: ALL_CAPS `MAP`

File: apps/photos/src/components/Sidebar/Preferences/index.tsx [hunk 6]
```diff
@@ -61,6 +72,11 @@ export default function Preferences({ open, onClose, onRootClose }) {
                 onClose={closeAdvancedSettings}
                 onRootClose={onRootClose}
             />
+            <AdvancedMapSettings
```
Comment (abhinavkgrd) [inline apps/photos/src/components/Sidebar/Preferences/index.tsx:75]: `<MapSettings>`

File: apps/photos/src/components/Sidebar/AdvancedMapSetting.tsx [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+import Titlebar from 'components/Titlebar';
+import { useEffect, useState } from 'react';
+import { t } from 'i18next';
+import { EnteMenuItem } from 'components/Menu/EnteMenuItem';
+import { MenuItemGroup } from 'components/Menu/MenuItemGroup';
+import MapSettings from './MapSettings';
+import { getData, LS_KEYS } from 'utils/storage/localStorage';
+
+export default function AdvancedMapSettings({ open, onClose, onRootClose }) {
+    const [mapSettingsView, setMapSettingsView] = useState(false);
```
Comment (abhinavkgrd) [inline apps/photos/src/components/Sidebar/AdvancedMapSetting.tsx:12]: `modifyMapEnabledSettingsView`

File: apps/photos/src/components/Sidebar/MapSettings.tsx [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+            const initialMapEnabled =
+                storedMapEnabled !== null ? storedMapEnabled.mapEnabled : false;
+            setMapEnabled(initialMapEnabled);
+        };
+        main();
+    }, []);
+
+    useEffect(() => {
+        setData(LS_KEYS.MAPENABLED, { mapEnabled });
+        updateMapEnabledStatus(mapEnabled);
```
Comment (abhinavkgrd) [inline apps/photos/src/components/Sidebar/MapSettings.tsx:28]: this should not be here but inside the `updateMapEnabled` function

File: apps/photos/src/components/Sidebar/MapSettings.tsx [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+    }, [mapEnabled]);
+
+    const updateMapEnabled = (enabled) => {
+        try {
+            setMapEnabled(enabled);
+        } catch (e) {
+            logError(e, 'Error while updating mapEnabled');
```
Comment (abhinavkgrd) [inline apps/photos/src/components/Sidebar/MapSettings.tsx:35]: you forgot to throw here

File: apps/photos/src/components/Sidebar/MapSettings.tsx [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+            logError(e, 'Enable Map failed');
+            somethingWentWrong();
+        }
+    };
+
+    const handleRootClose = () => {
+        onClose();
+        onRootClose();
+    };
+    const handleDrawerClose: DialogProps['onClose'] = (_, reason) => {
```
Comment (abhinavkgrd) [inline apps/photos/src/components/Sidebar/MapSettings.tsx:62]: try to add empty line after each function declaration

File: apps/photos/src/components/Sidebar/MapSettings.tsx [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+        }
+    };
+    return (
+        <Box>
+            <EnteDrawer
+                anchor="left"
+                transitionDuration={0}
+                open={open}
+                onClose={handleDrawerClose}
+                BackdropProps={{
```
Comment (abhinavkgrd) [inline apps/photos/src/components/Sidebar/MapSettings.tsx:76]: This is depreciated property
lets update this

File: apps/photos/src/components/Sidebar/EnableMap.tsx [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+                    <Typography color="text.muted">
+                        <Trans i18nKey={'ENABLE_MAP_DESCRIPTION_1'} />
+                        <br />
+                        <br />
+
+                        <Trans i18nKey={'ENABLE_MAP_DESCRIPTION_2'} />
+                        <br />
+                        <br />
+
+                        <Trans i18nKey={'ENABLE_MAP_DESCRIPTION_3'} />
```
Comment (abhinavkgrd) [inline apps/photos/src/components/Sidebar/EnableMap.tsx:26]: can we convert this to https://github.com/ente-io/photos-web/blob/6514198cf310b3c9a1c190dd92a8be6efd51a41d/apps/photos/src/components/MachineLearning/MLSearchSettings/enableFaceSearch.tsx#L57-L72

File: apps/photos/src/components/Sidebar/DisableMap.tsx [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+                onRootClose={onRootClose}
+            />
+            <Stack py={'20px'} px={'8px'} spacing={'32px'}>
+                <Box px={'8px'}>
+                    {' '}
+                    <Typography color="text.muted">
+                        <Trans i18nKey={'DISABLE_MAP_DESCRIPTION_1'} />
+                        <br />
+                        <br />
+                        <Trans i18nKey={'DISABLE_MAP_DESCRIPTION_2'} />
```
Comment (abhinavkgrd) [inline apps/photos/src/components/Sidebar/DisableMap.tsx:21]: same as enableMap comment

File: apps/photos/src/components/Sidebar/AdvancedMapSetting.tsx [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+        if (reason === 'backdropClick') {
+            handleRootClose();
+        } else {
+            onClose();
+        }
+    };
+    useEffect(() => {
+        const mapEnabledValue = getData(LS_KEYS.MAPENABLED);
+        setMapEnabledToggle(mapEnabledValue.mapEnabled);
+    }, [mapSettingsView]);
```
Comment (abhinavkgrd) [inline apps/photos/src/components/Sidebar/AdvancedMapSetting.tsx:33]: this should called only on load no?

File: apps/photos/src/pages/gallery/index.tsx [hunk 15]
```diff
@@ -357,6 +358,15 @@ export default function Gallery() {
         }
     }, [router.isReady]);
 
+    useEffect(() => {
+        const main = async () => {
+            const remoteMapValue = await getMapEnabledStatus();
```
Comment (abhinavkgrd) [inline apps/photos/src/pages/gallery/index.tsx:363]: lets keep this inside the single `syncWithRemote` function

File: apps/photos/src/utils/storage/localStorage.ts [hunk 16]
```diff
@@ -22,6 +22,7 @@ export enum LS_KEYS {
     WAIT_TIME = 'waitTime',
     API_ENDPOINT = 'apiEndpoint',
     LOCALE = 'locale',
+    MAPENABLED = 'mapEnabled',
```
Comment (abhinavkgrd) [inline apps/photos/src/utils/storage/localStorage.ts:25]: `MAP_ENABLED`

**Pull request merged**